File: internal/auth/repository.go
package auth

import (
	"context"
	"errors"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"

	"qwikkle-api/internal/db"
	"qwikkle-api/internal/types"
)

var (
	ErrIdentityTaken      = errors.New("identity already registered")
	ErrInvalidCredentials = errors.New("invalid qkId or password")
	ErrUserNotFound       = errors.New("user not found")
	ErrSessionNotFound    = errors.New("session not found")
)

type Repository interface {
	CreateUser(ctx context.Context, in CreateUserInput) (*User, error)
	GetUserByID(ctx context.Context, id string) (*User, error)
	GetUserByQKID(ctx context.Context, qkID string) (*User, error)
	UpdateUserProfile(ctx context.Context, userID string, in UpdateUserProfileInput) (*User, error)
	CreateSession(ctx context.Context, userID string, refreshTokenHash string, expiresAt time.Time, userAgent string, ip string) (string, error)
	GetSessionByRefreshTokenHash(ctx context.Context, refreshTokenHash string) (*Session, error)
	RotateSession(ctx context.Context, sessionID string, refreshTokenHash string, expiresAt time.Time) error
	RevokeSession(ctx context.Context, sessionID string) error
}

type Session struct {
	ID               string
	UserID           string
	RefreshTokenHash string
	CreatedAt        time.Time
	ExpiresAt        time.Time
	RevokedAt        *time.Time
	UserAgent        *string
	IP               *string
}

type postgresRepo struct {
	pool *db.Pool
}

func NewPostgresRepository(pool *db.Pool) Repository {
	return &postgresRepo{pool: pool}
}

type CreateUserInput struct {
	QKID              string
	Email             *string
	PasswordHash      string
	Role              string
	Status            types.AccountStatus
	FirstName         *string
	LastName          *string
	Phone             *string
	AvatarURL         *string
	Gender            *string
	DateOfBirth       *time.Time
	Country           *string
	Interests         []string
	AvatarStorageKey  *string
	AvatarDownloadURL *string
}

type UpdateUserProfileInput struct {
	Email             *string
	FirstName         *string
	LastName          *string
	Phone             *string
	AvatarURL         *string
	Gender            *string
	DateOfBirth       *time.Time
	Country           *string
	Interests         *[]string
	AvatarStorageKey  *string
	AvatarDownloadURL *string
}

func (r *postgresRepo) CreateUser(ctx context.Context, in CreateUserInput) (*User, error) {
	interests := in.Interests
	if interests == nil {
		interests = []string{}
	}

	const q = `
		INSERT INTO users (
			qk_id,
			email,
			password_hash,
			role,
			status,
			first_name,
			last_name,
			phone,
			avatar_url,
			gender,
			date_of_birth,
			country,
			interests,
			avatar_storage_key,
			avatar_download_url
		)
		VALUES (
			$1,
			$2,
			$3,
			$4,
			$5,
			NULLIF($6, ''),
			NULLIF($7, ''),
			NULLIF($8, ''),
			NULLIF($9, ''),
			NULLIF($10, ''),
			$11,
			NULLIF($12, ''),
			$13,
			NULLIF($14, ''),
			NULLIF($15, '')
		)
		RETURNING
			id,
			qk_id,
			email,
			password_hash,
			role,
			status,
			first_name,
			last_name,
			phone,
			avatar_url,
			gender,
			date_of_birth,
			country,
			interests,
			avatar_storage_key,
			avatar_download_url,
			organization_id,
			created_at,
			last_login_at
	`

	var u User
	err := r.pool.QueryRow(
		ctx,
		q,
		in.QKID,
		in.Email,
		in.PasswordHash,
		in.Role,
		in.Status,
		in.FirstName,
		in.LastName,
		in.Phone,
		in.AvatarURL,
		in.Gender,
		in.DateOfBirth,
		in.Country,
		interests,
		in.AvatarStorageKey,
		in.AvatarDownloadURL,
	).Scan(
		&u.ID,
		&u.QKID,
		&u.Email,
		&u.PasswordHash,
		&u.Role,
		&u.Status,
		&u.FirstName,
		&u.LastName,
		&u.Phone,
		&u.AvatarURL,
		&u.Gender,
		&u.DateOfBirth,
		&u.Country,
		&u.Interests,
		&u.AvatarStorageKey,
		&u.AvatarDownloadURL,
		&u.OrganizationID,
		&u.CreatedAt,
		&u.LastLoginAt,
	)
	if err != nil {
		var pgErr *pgconn.PgError
		if errors.As(err, &pgErr) {
			if pgErr.ConstraintName == "users_email_key" || pgErr.ConstraintName == "users_qk_id_key" {
				return nil, ErrIdentityTaken
			}
		}
		return nil, err
	}
	return &u, nil
}

func (r *postgresRepo) GetUserByQKID(ctx context.Context, qkID string) (*User, error) {
	const q = `
		SELECT
			id,
			qk_id,
			email,
			password_hash,
			role,
			status,
			first_name,
			last_name,
			phone,
			avatar_url,
			gender,
			date_of_birth,
			country,
			interests,
			avatar_storage_key,
			avatar_download_url,
			organization_id,
			created_at,
			last_login_at
		FROM users
		WHERE qk_id = $1
	`

	var u User
	err := r.pool.QueryRow(ctx, q, qkID).Scan(
		&u.ID,
		&u.QKID,
		&u.Email,
		&u.PasswordHash,
		&u.Role,
		&u.Status,
		&u.FirstName,
		&u.LastName,
		&u.Phone,
		&u.AvatarURL,
		&u.Gender,
		&u.DateOfBirth,
		&u.Country,
		&u.Interests,
		&u.AvatarStorageKey,
		&u.AvatarDownloadURL,
		&u.OrganizationID,
		&u.CreatedAt,
		&u.LastLoginAt,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, ErrUserNotFound
		}
		return nil, err
	}
	return &u, nil
}

func (r *postgresRepo) GetUserByID(ctx context.Context, id string) (*User, error) {
	const q = `
		SELECT
			id,
			qk_id,
			email,
			password_hash,
			role,
			status,
			first_name,
			last_name,
			phone,
			avatar_url,
			gender,
			date_of_birth,
			country,
			interests,
			avatar_storage_key,
			avatar_download_url,
			organization_id,
			created_at,
			last_login_at
		FROM users
		WHERE id = $1
	`

	var u User
	err := r.pool.QueryRow(ctx, q, id).Scan(
		&u.ID,
		&u.QKID,
		&u.Email,
		&u.PasswordHash,
		&u.Role,
		&u.Status,
		&u.FirstName,
		&u.LastName,
		&u.Phone,
		&u.AvatarURL,
		&u.Gender,
		&u.DateOfBirth,
		&u.Country,
		&u.Interests,
		&u.AvatarStorageKey,
		&u.AvatarDownloadURL,
		&u.OrganizationID,
		&u.CreatedAt,
		&u.LastLoginAt,
	)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, ErrUserNotFound
		}
		return nil, err
	}
	return &u, nil
}

func (r *postgresRepo) UpdateUserProfile(ctx context.Context, userID string, in UpdateUserProfileInput) (*User, error) {
	const q = `
		UPDATE users
		SET
			email = COALESCE($2, email),
			first_name = COALESCE(NULLIF($3, ''), first_name),
			last_name = COALESCE(NULLIF($4, ''), last_name),
			phone = COALESCE(NULLIF($5, ''), phone),
			avatar_url = COALESCE(NULLIF($6, ''), avatar_url),
			gender = COALESCE(NULLIF($7, ''), gender),
			date_of_birth = COALESCE($8, date_of_birth),
			country = COALESCE(NULLIF($9, ''), country),
			interests = COALESCE($10, interests),
			avatar_storage_key = COALESCE(NULLIF($11, ''), avatar_storage_key),
			avatar_download_url = COALESCE(NULLIF($12, ''), avatar_download_url)
		WHERE id = $1
	`
	_, err := r.pool.Exec(
		ctx,
		q,
		userID,
		in.Email,
		in.FirstName,
		in.LastName,
		in.Phone,
		in.AvatarURL,
		in.Gender,
		in.DateOfBirth,
		in.Country,
		in.Interests,
		in.AvatarStorageKey,
		in.AvatarDownloadURL,
	)
	if err != nil {
		return nil, err
	}
	return r.GetUserByID(ctx, userID)
}

func (r *postgresRepo) CreateSession(
	ctx context.Context,
	userID string,
	refreshTokenHash string,
	expiresAt time.Time,
	userAgent string,
	ip string,
) (string, error) {
	const q = `
		INSERT INTO sessions (user_id, refresh_token_hash, expires_at, user_agent, ip)
		VALUES ($1, $2, $3, NULLIF($4, ''), NULLIF($5, '')::inet)
		RETURNING id
	`

	var id string
	if err := r.pool.QueryRow(ctx, q, userID, refreshTokenHash, expiresAt, userAgent, ip).Scan(&id); err != nil {
		return "", err
	}
	return id, nil
}

func (r *postgresRepo) GetSessionByRefreshTokenHash(
	ctx context.Context,
	refreshTokenHash string,
) (*Session, error) {
	const q = `
		SELECT id, user_id, refresh_token_hash, created_at, expires_at, revoked_at, user_agent, ip::text
		FROM sessions
		WHERE refresh_token_hash = $1
	`

	var s Session
	if err := r.pool.QueryRow(ctx, q, refreshTokenHash).Scan(
		&s.ID,
		&s.UserID,
		&s.RefreshTokenHash,
		&s.CreatedAt,
		&s.ExpiresAt,
		&s.RevokedAt,
		&s.UserAgent,
		&s.IP,
	); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, ErrSessionNotFound
		}
		return nil, err
	}

	return &s, nil
}

func (r *postgresRepo) RotateSession(ctx context.Context, sessionID string, refreshTokenHash string, expiresAt time.Time) error {
	const q = `
		UPDATE sessions
		SET refresh_token_hash = $1, expires_at = $2
		WHERE id = $3 AND revoked_at IS NULL
	`

	ct, err := r.pool.Exec(ctx, q, refreshTokenHash, expiresAt, sessionID)
	if err != nil {
		return err
	}
	if ct.RowsAffected() == 0 {
		return ErrSessionNotFound
	}
	return nil
}

func (r *postgresRepo) RevokeSession(ctx context.Context, sessionID string) error {
	const q = `
		UPDATE sessions
		SET revoked_at = NOW()
		WHERE id = $1 AND revoked_at IS NULL
	`

	ct, err := r.pool.Exec(ctx, q, sessionID)
	if err != nil {
		return err
	}
	if ct.RowsAffected() == 0 {
		return ErrSessionNotFound
	}
	return nil
}

File: src/auth/repository.ts
import type { Sql } from 'postgres';
import type { User, Session, SignupInput, LoginInput, UpdateUserProfileInput } from '../types';
import { generateId, hashPassword, hashToken } from '../utils/crypto';
import { normalizeQKID } from '../utils/validation';

export interface AuthRepository {
  getUserByQKID(qkId: string): Promise<User | null>;
  getUserByID(id: string): Promise<User | null>;
  createUser(input: SignupInput): Promise<User>;
  updateUserProfile(id: string, input: UpdateUserProfileInput): Promise<User>;
  createSession(userId: string, refreshTokenHash: string, expiresAt: Date, userAgent?: string, ipAddress?: string): Promise<Session>;
  getSessionByRefreshTokenHash(refreshTokenHash: string): Promise<Session | null>;
  rotateSession(sessionId: string, newRefreshTokenHash: string, expiresAt: Date): Promise<void>;
  revokeSession(sessionId: string): Promise<void>;
  bootstrapAdmin(): Promise<void>;
}

export class PostgresAuthRepository implements AuthRepository {
  constructor(private sql: Sql) {}

  async getUserByQKID(qkId: string): Promise<User | null> {
    const normalizedQKID = normalizeQKID(qkId);
    const rows = await this.sql`
      SELECT * FROM users WHERE qk_id = ${normalizedQKID}
    `;
    
    if (rows.length === 0) return null;
    return this.mapRowToUser(rows[0]);
  }

  async getUserByID(id: string): Promise<User | null> {
    const rows = await this.sql`
      SELECT * FROM users WHERE id = ${id}
    `;
    
    if (rows.length === 0) return null;
    return this.mapRowToUser(rows[0]);
  }

  async createUser(input: SignupInput): Promise<User> {
    const id = generateId();
    const normalizedQKID = normalizeQKID(input.qkId);
    const passwordHash = await hashPassword(input.password);
    
    const rows = await this.sql`
      INSERT INTO users (
        id, qk_id, email, password_hash, first_name, last_name, phone,
        gender, date_of_birth, country, interests, avatar_url,
        avatar_storage_key, avatar_download_url, role, status, created_at, updated_at
      ) VALUES (
        ${id}, ${normalizedQKID}, ${input.email || null}, ${passwordHash},
        ${input.firstName || null}, ${input.lastName || null}, ${input.phone || null},
        ${input.gender || null}, ${input.dateOfBirth || null}, ${input.country || null},
        ${JSON.stringify(input.interests)}, ${null}, ${null}, ${null},
        'user', 'active', NOW(), NOW()
      ) RETURNING *
    `;
    
    return this.mapRowToUser(rows[0]);
  }

  async updateUserProfile(id: string, input: UpdateUserProfileInput): Promise<User> {
    const rows = await this.sql`
      UPDATE users SET
        email = ${input.email || null},
        first_name = ${input.firstName || null},
        last_name = ${input.lastName || null},
        phone = ${input.phone || null},
        gender = ${input.gender || null},
        date_of_birth = ${input.dateOfBirth || null},
        country = ${input.country || null},
        interests = ${input.interests ? JSON.stringify(input.interests) : null},
        updated_at = NOW()
      WHERE id = ${id}
      RETURNING *
    `;
    
    if (rows.length === 0) throw new Error('User not found');
    return this.mapRowToUser(rows[0]);
  }

  async createSession(
    userId: string,
    refreshTokenHash: string,
    expiresAt: Date,
    userAgent?: string,
    ipAddress?: string
  ): Promise<Session> {
    const id = generateId();
    
    const rows = await this.sql`
      INSERT INTO sessions (
        id, user_id, refresh_token_hash, expires_at, created_at,
        user_agent, ip_address
      ) VALUES (
        ${id}, ${userId}, ${refreshTokenHash}, ${expiresAt}, NOW(),
        ${userAgent || null}, ${ipAddress || null}
      ) RETURNING *
    `;
    
    return this.mapRowToSession(rows[0]);
  }

  async getSessionByRefreshTokenHash(refreshTokenHash: string): Promise<Session | null> {
    const rows = await this.sql`
      SELECT * FROM sessions WHERE refresh_token_hash = ${refreshTokenHash}
    `;
    
    if (rows.length === 0) return null;
    return this.mapRowToSession(rows[0]);
  }

  async rotateSession(sessionId: string, newRefreshTokenHash: string, expiresAt: Date): Promise<void> {
    await this.sql`
      UPDATE sessions SET
        refresh_token_hash = ${newRefreshTokenHash},
        expires_at = ${expiresAt},
        updated_at = NOW()
      WHERE id = ${sessionId}
    `;
  }

  async revokeSession(sessionId: string): Promise<void> {
    await this.sql`
      UPDATE sessions SET revoked_at = NOW() WHERE id = ${sessionId}
    `;
  }

  async bootstrapAdmin(): Promise<void> {
    const adminExists = await this.sql`
      SELECT 1 FROM users WHERE role = 'admin' LIMIT 1
    `;
    
    if (adminExists.length > 0) return;
    
    const adminId = generateId();
    const passwordHash = await hashPassword('admin123');
    
    await this.sql`
      INSERT INTO users (
        id, qk_id, email, password_hash, role, status, created_at, updated_at
      ) VALUES (
        ${adminId}, 'admin', 'admin@qwikkle.local', ${passwordHash},
        'admin', 'active', NOW(), NOW()
      )
    `;
  }

  private mapRowToUser(row: any): User {
    return {
      id: row.id,
      qkId: row.qk_id,
      email: row.email,
      passwordHash: row.password_hash,
      firstName: row.first_name,
      lastName: row.last_name,
      phone: row.phone,
      gender: row.gender,
      dateOfBirth: row.date_of_birth,
      country: row.country,
      interests: row.interests ? JSON.parse(row.interests) : [],
      avatarUrl: row.avatar_url,
      avatarStorageKey: row.avatar_storage_key,
      avatarDownloadUrl: row.avatar_download_url,
      role: row.role,
      status: row.status,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
      lastLoginAt: row.last_login_at,
    };
  }

  private mapRowToSession(row: any): Session {
    return {
      id: row.id,
      userId: row.user_id,
      refreshTokenHash: row.refresh_token_hash,
      expiresAt: row.expires_at,
      createdAt: row.created_at,
      revokedAt: row.revoked_at,
      userAgent: row.user_agent,
      ipAddress: row.ip_address,
    };
  }
}
